pkg/karmadactl/logs.go

```diff
@@ -112,17 +114,11 @@ func (o *LogsOptions) Complete(karmadaConfig KarmadaConfig, cmd *cobra.Command,
 		return cmdutil.UsageErrorf(cmd, "%s", logsUsageErrStr)
 	}
 
-	karmadaRestConfig, err := karmadaConfig.GetRestConfig(o.KarmadaContext, o.KubeConfig)
-	if err != nil {
-		return fmt.Errorf("failed to get control plane rest config. context: %s, kube-config: %s, error: %v",
-			o.KarmadaContext, o.KubeConfig, err)
-	}
-	clusterInfo, err := getClusterInfo(karmadaRestConfig, o.Cluster, o.KubeConfig, o.KarmadaContext)
+	memberFactory, err := f.FactoryForMemberCluster(o.Cluster)
 	if err != nil {
 		return err
 	}
-	f := getFactory(o.Cluster, clusterInfo, o.Namespace)
-	return o.KubectlLogsOptions.Complete(f, cmd, args)
```

Comment:
> At present, there are many places in karmadactl, which are building clients and factories repeatedly, such as init, get, log, exec, etc.

Is it you mean building clients and factories repeatedly?

Reply:
@RainbowMango 

Currently, each subcommand has the logic to create client and cmdutil.Factory, for example:

- https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/apply.go#L114
- https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/cordon.go#L223
- https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/deinit.go#L92
- https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/describe.go#L105
- https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/exec.go#L114
- https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/get.go#L294
- https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/join.go#L161
- https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/join.go#L242
- https://github.com/karmada-io/karmada/blob/master/pkg/karmadactl/promote.go#L207

I expect each sub command take factory as input parameter which factory provides functions to create kube client, karmada client  and cmdutil.Factory for member cluster.

```go
	groups := templates.CommandGroups{
		{
			Message: "Basic Commands:",
			Commands: []*cobra.Command{
				NewCmdGet(factory, parentCommand, ioStreams),
			},
		},
		{
			Message: "Cluster Registeration Commands:",
			Commands: []*cobra.Command{
				cmdinit.NewCmdInit(factory, parentCommand),
				NewCmdDeInit(factory, parentCommand),
				addons.NewCommandAddons(factory, parentCommand),
				NewCmdJoin(factory, parentCommand),
				NewCmdUnjoin(factory, parentCommand),
			},
		},
		{
			Message: "Cluster Management Commands:",
			Commands: []*cobra.Command{
				NewCmdCordon(factory, parentCommand),
				NewCmdUncordon(factory, parentCommand),
				NewCmdTaint(factory, parentCommand),
			},
		},
		{
			Message: "Troubleshooting and Debugging Commands:",
			Commands: []*cobra.Command{
				NewCmdLogs(factory, parentCommand, ioStreams),
				NewCmdExec(factory, parentCommand, ioStreams),
				NewCmdDescribe(factory, parentCommand, ioStreams),
			},
		},
		{
			Message: "Advanced Commands:",
			Commands: []*cobra.Command{
				NewCmdApply(factory, parentCommand, ioStreams),
				NewCmdPromote(factory, parentCommand),
			},
		},
	}
```

Reply:
There seems not all subcommands needing `karmadaclientset`, should we need to provide a common `kubeclientset` in my `Factory`?

Reply:
Factory extends cmdutil.Factory, so kubeclient can be created by this factory.

https://github.com/karmada-io/karmada/pull/2202/files#diff-8aa0689c77923337c7ccd21b21fec3fd92f16fee1c417149d6fd89d4b2af45d0R33

Reply:
> Factory extends cmdutil.Factory, so kubeclient can be created by this factory.

How to do it for `deinit` subcommand.

Reply:
@lonelyCZ 

you are right.

we need provide a common function for  `init`, `deinit` or `join` command to talk with those cluster which are not managed by karmada.

I'm not sure that we need to provide a common kubeclientset in my Factory. because my  Factory is designed for talking with karmada APIServer.

Reply:
If we want to provide a common kubeclientset in my Factory, how about defining it like this?

```go
type Factory interface {
	cmdutil.Factory

	// KarmadaClientSet returns a karmada clientset
	KarmadaClientSet() (karmadaclientset.Interface, error)

	// FactoryForMemberCluster returns a cmdutil.Factory for the member cluster
	FactoryForMemberCluster(clusterName string) (cmdutil.Factory, error)

	// KubernetesClientForCluster returns a kubernetes clientset for a cluster. 
	// If kubeconfig is not empty, it will use the kubeconfig to create the clientset. Otherwise, it will use the config file used by this factory.
	KubernetesClientForCluster(kubeconfig, context string) (kubernetes.Interface, error)
}
```

Reply:
How to directly use `KubernetesClientSet(kubeconfig, context string)`?

But I am worried about that we init the unuseful `KarmadaClientSet` for `init`, `deinit` commends. Perhaps, we can don't care these cmds firstly, not need to pass `f` to them.

Reply:
In future, the adm and control commands possibly will be seprated, like `kubectl` and `kubeadm`

Reply:
> In future, the adm and control commands possibly will be seprated, like `kubectl` and `kubeadm`

I like it.

> How to directly use `KubernetesClientSet(kubeconfig, context string)`?
> 
> But I am worried about that we init the unuseful `KarmadaClientSet` for `init`, `deinit` commends. Perhaps, we can don't care these cmds firstly, not need to pass `f` to them.

 It makes sense for me.

Reply:
@lonelyCZ  Does this PR need to be changed?

Reply:
It should be no problem, but there seem a lot of works at #2349 , I'm afraid I won't have enough time to review it, because I need to develop #2282 for coming new version these days. 

Since there are several new commands to be added, could we postpone this cleanup after the new release? @carlory

Reply:
@lonelyCZ ok